### PR TITLE
Reduce allocations

### PR DESCRIPTION
**Pull request by mkaminski1988**
@crocodele In this branch AST nodes are retrieved from a stack-allocated array instead of dynamic allocation. When I run the benchmarks on the branch, the `JsonPathPhpExtensionBenchmark` beats `NativePhpBenchmark` on `benchMediumDataset`. Do you get the same results?

### CODE REVIEW

**Comment by mkaminski1988** [comment]
Would having opcache enabled close the performance gap between the implementations, thus rendering the extension moot? IIRC we would need to run this with PHP-FPM to test this out. (It's been like 5 years since I last developed with PHP, so I might be wrong.)

**Comment by crocodele** [comment]
> When I run the benchmarks on the branch, the `JsonPathPhpExtensionBenchmark` beats `NativePhpBenchmark` on `benchMediumDataset`. Do you get the same results?

Hmm, no, I get pretty much the same numbers on `refactor/reduce_alloc` as on `main`. :thinking: 

The `NativePhpBenchmark` performance seems to depend on the system though to some extent. On my main machine I get values around 240 microseconds for `benchMediumDataset`, on another machine around 360 microseconds, which is slightly less than `JsonPathPhpExtensionBenchmark` on both machines.

> Would having opcache enabled close the performance gap between the implementations, thus rendering the extension moot?

That's a good point. I ran the benchmarks again with `opcache.enable_cli = 1`. In the medium benchmark it cut 45% and 15% off the execution time for the two tested PHP libraries. The extension is still a lot faster though. No improvement from OPCache in the tiny or huge benchmarks.

**Comment by mkaminski1988** [comment]
> The NativePhpBenchmark performance seems to depend on the system though to some extent. On my main machine I get values around 240 microseconds for benchMediumDataset, on another machine around 360 microseconds, which is slightly less than JsonPathPhpExtensionBenchmark on both machines.

Interesting, did you try with both PHP 7.4 and 8?

**Comment by mkaminski1988** [comment]
@crocodele I reverted the buggy dynamic memory pool. See commit message in dbce73bbcadbe31cc008ef62f93a642661ae1262 for a summary of the changes.

**Comment by mkaminski1988** [comment]
@crocodele I reverted the buggy dynamic memory pool. See commit message in dbce73bbcadbe31cc008ef62f93a642661ae1262 for a summary of the changes.

**Comment by mkaminski1988** [comment]
@crocodele I reverted the buggy dynamic memory pool. See commit message in 7a313cd25a85569d4cfa5e8ef3b101a9be802d9a for a summary of the changes.

File: src/jsonpath/parser.c [hunk 1]
```diff
@@ -147,14 +154,13 @@ static bool parse_filter_list(PARSER_PARAMS, struct ast_node* tok) {
         zend_throw_exception(spl_ce_RuntimeException, "Array slice indexes must be integers", 0);
         return false;
```
Comment (crocodele) [inline src/jsonpath/parser.c:155]: I get a memory leak in tests/015.phpt (Test array slice with strings as start, end, and step), which I managed to fix in the node pool PR with:

```
if (sep_found == AST_INDEX_SLICE) {
  zend_throw_exception(spl_ce_RuntimeException, "Array slice indexes must be integers", 0);

  /* If we end up here, we may already have parsed one or more string literals into the */
  /* hash table under the assumption that we were dealing with an index filter. Free those. */
  if (tok->data.d_nodes.ht != NULL) {
    zend_array_destroy(tok->data.d_nodes.ht);
    tok->data.d_nodes.count = 0;
  }

  return false;
}
```

Something like that perhaps?

**Review by crocodele — CHANGES_REQUESTED** [review]

File: src/jsonpath/parser.c (hunk 1 above)
Comment (mkaminski1988) [reply src/jsonpath/parser.c:163]: Thanks for flagging! I made a go at unifying the data structure for slices, index lists, and node selector lists using a HashTable, which incidentally solve this leak. Check out fd756e7560d4815ed78fcc03ad2a382c03751dd0.

**Review by crocodele — APPROVED** [review]

**Pull request merged**